### PR TITLE
CORE-1905: Validate PublishRequests with StreamMessageValidator

### PR DESCRIPTION
**Pull request by hpihkala**
**Main change:**

Use the new [`StreamMessageValidator`](https://github.com/streamr-dev/streamr-client-protocol-js/blob/master/src/utils/StreamMessageValidator.js) abstraction to validate `PublishRequest`s. 

The above achieves the following:
- Upgrade to the [type-specific message validation logic](https://github.com/streamr-dev/streamr-specs/blob/master/validation.md) which cover also key exchange requests & responses
- Remove code duplication in broker and JS client by using a shared abstraction
- This is one important step towards trustlessness: signed messages can now be validated by _whoever observes them_ (from broker point of view, this means the centralized `sessionToken` is no longer needed to validate `PublishRequest`s.

**Other minor changes:**
- Cleanup a few code sections by refactoring promise handling from `.then()` to `async/await`
- Change MQTT topic-to-stream mapping logic to `topic == streamId`

**Remaining concerns**

The broker now validates the signatures of messages, whereas previously it did not. Validating signatures may introduce a significant CPU load. This needs some further testing and benchmarking. 

In our future goal network, doing cryptographic validation is unavoidable, but there are some ideas for (temporary or permanent) optimizations that can be done:
- The signature validation is done with whatever implementation `ethers.js` uses. Faster implementations might exist.
- It's possible for the broker to _not_ validate all messages. Subscribers will validate all messages anyway, so brokers _also_ validating messages is a trade-off between broker CPU usage vs. the desired network-level functionality of stopping spoofed messages early. A potential sweet spot could be that the broker only validates the _first message_ for each `[connection, publisherId]` pair to establish that the client holds the key for `publisherId`, and thereafter skip signature validation for subsequent messages for the same `publisherId`.

**TODOs**

- ~~Benchmark validation performance~~
- ~~Decide on signature validation strategy (all messages vs. first message vs. no messages)~~
- ~~Implement chosen signature validation strategy~~
- ~~Import updated protocol library~~

### CODE REVIEW

**Comment by hpihkala** [comment]
I did some benchmarking of the signature validation with `ethers.utils.verifyMessage(...)` (or whatever it uses internally). It turns out to be extremely slow: I get around **4.1 messages/second** on my laptop. Yes, that's right!

So this will definitely need **both** a faster library **and** a workaround to avoid validating all of the messages at the broker.

File: src/websocket/FieldDetector.js [hunk 1]
```diff
@@ -4,7 +4,9 @@ module.exports = class FieldDetector {
         this.configuredStreamIds = new Set()
     }
 
-    async detectAndSetFields(stream, streamMessage, apiKey, sessionToken) {
+    async detectAndSetFields(streamMessage, apiKey, sessionToken) {
+        const stream = await this.streamFetcher.fetch(streamMessage.getStreamId(), apiKey, sessionToken)
```
Comment (mirpo) [inline src/websocket/FieldDetector.js:8]: as I remember this call was super slow, that's why we were caching this in `handlePublishRequest`.

**Comment by hpihkala** [comment]
Ok I have an implementation of signature verification, which is 1000x faster than `ethers.js` and `web3.js`. Benchmark shows around 4k verifications/sec. It's using native code. Not sure how to improve performance in browser yet, but this looks pretty good for broker.

The code is currently in the protocol repo. Next I'll see if I can package it reasonably in there (so that it works both on node and browser, or do I have to move it to `broker`.

**Comment by mirpo** [comment]
Good news about performance! In browser we can try `wasm` with fallback to `ethers.js`.

**Comment by hpihkala** [comment]
> Good news about performance! In browser we can try `wasm` with fallback to `ethers.js`.

That's interesting, could definitely try. Let's return to that topic once we have the `StreamMessageValidator` in the JS client as well. First step there is to bring in the abstraction (and continue to use `ethers.js`), second step would be to optimize performance.

**Comment by hpihkala** [comment]
The performance improvements for node.js are opened in [this PR](https://github.com/streamr-dev/streamr-client-protocol-js/pull/35). Once merged, I'll bring the new version into this PR. That should remove the performance obstacles for the time being, as the improved code benchmarks around 3500+ verifications/sec.

**Comment by hpihkala** [comment]
I'm gonna have to upgrade to new major version `5.x` of the protocol package (which adds support for ControlLayer `V2`), because that's what the performance improvements got added against.

Would have liked to ideally do those two things as separate PRs, but it's difficult to separate those changes in the protocol repo anymore. So I'll just run with it.

**Comment by hpihkala** [comment]
Status update:
- Looks like `StreamMessage` objects are being passed between `broker` and `network` layers. This means that I will have to update `network` to use the new protocol version first, and then return to `broker` to finish up.
- Also noticed unit test file `_WebsocketServer._test.js` which is not being run at all. Asked Eric & Miroslav about it, but I didn't do anything about it yet.

So this remains WIP for now.

**Comment by hpihkala** [comment]
Status update: This PR is quite ready to be merged, but I'm waiting for the current `master` (containing the DU/permissions changes) to be deployed before merging this, just to avoid pushing too many changes into one deployment.

File: src/websocket/FieldDetector.js (hunk 1 above)
Comment (hpihkala) [reply src/websocket/FieldDetector.js]: @mirpo `streamFetcher.fetch` already caches the results (with memoize), so to me it seemed that it was basically being double-cached for no obvious reason. Please let me know if I missed something!

**Comment by hpihkala** [comment]
Ok, was holding this back until the permission changes were deployed, and now they are. I've been exercising this branch with Melchior's cross-client tests and some other things, and it seems ok. Ready to merge if tests pass in CI, waiting for them to complete.

**Comment by hpihkala** [comment]
Tests pass but jest occasionally segfaults after reporting integration test results. 🤷🏻

```
Test Suites: 13 passed, 13 total
2934Tests:       80 passed, 80 total
2935Snapshots:   0 total
2936Time:        134.771 s
2937Ran all test suites matching /test\/integration/i.
2938Segmentation fault (core dumped)
2939npm ERR! code ELIFECYCLE
2940npm ERR! errno 139
2941npm ERR! streamr-broker@10.0.0 test-integration: `jest --detectOpenHandles --forceExit test/integration`
```

**Pull request merged**